quipucords/api/credential/serializer.py

```diff
@@ -235,3 +243,64 @@ def prepare_data_for_update(cls, data: dict):
             data["ssh_keyfile"] = None
         elif ssh_keyfile:
             data["password"] = None
+
+
+class AuthTokenOrUserPassSerializer(CredentialSerializer):
+    """Serialize credentials that require username+password or auth token."""
+
+    class Meta:
+        """Serializer configuration."""
+
+        model = Credential
+        fields = [
+            "cred_type",
+            "id",
+            "name",
+            "password",
+            "username",
+            "auth_token",
+            "sources",
+        ]
+        extra_kwargs = {
+            "password": {**ENCRYPTED_FIELD_KWARGS},
+            "auth_token": {**ENCRYPTED_FIELD_KWARGS},
+        }
+
+    def validate(self, attrs):
+        """Run validations that require more than one field."""
+        has_user_or_pass, has_auth_token = self._get_user_pass_or_auth_token(attrs)
+
+        if not has_auth_token ^ has_user_or_pass:
```

Comment:
Minor quibble. This doesn't feel legible to me at a glance, probably because when I see `^`, I think of bitwise operations, and I have to pause to consider if these variables are more than just bools. I think I would prefer a more verbose, though arguably redundant, form of this check.

Reply:
mmm I think operator has an explicit xor function. would that help?

Reply:
```python
if not xor(has_auth_token, has_user_pass)
```

Reply:
or would u prefer it trully explicit like
```python
        if has_auth_token and has_user_or_pass or (not has_auth_token and not has_user_or_pass):

```

Reply:
The logic would be simplified if we had 2 ValidationErrors:

```
    if not has_auth_token and not has_user_or_pass:
        raise ValidationError("Missing authentication")

    if has_auth_token and has_user_or_pass:
        raise ValidationError("Either auth_token or username+password can be specified"):
```

first message could also be "Either auth_token or username_password must be specified"